src/Inciter/Carrier.C

```diff
@@ -244,9 +244,57 @@ Carrier::vol()
 }
 
 void
+Carrier::comvol( const std::vector< std::size_t >& gid,
+                 const std::vector< tk::real >& V )
+// *****************************************************************************
+//  Receive nodal volumes on chare-boundaries
+//! \param[in] gid Global mesh node IDs at which we receive volume contributions
+//! \param[in] V Partial sums of nodal volume contributions to chare-boundary
+//!   nodes
+//! \details This function receives contributions to m_vol, which stores the
+//!   nodal volumes. While m_vol stores own contributions, m_volc collects the
+//!   neighbor chare contributions during communication. This way work on m_vol
+//!   and m_volc is overlapped. The two are combined in totalvol().
+// *****************************************************************************
+{
+  Assert( V.size() == gid.size(), "Size mismatch" );
+
+  for (std::size_t i=0; i<gid.size(); ++i) {
+    auto bid = tk::cref_find( m_bid, gid[i] );
+    Assert( bid < m_volc.size(), "Indexing out of bounds" );
+    m_volc[ bid ] += V[i];
+  }
+
+  if (++m_nvol == m_msum.size()) {
+    m_nvol = 0;
+    contribute(
+       CkCallback(CkReductionTarget(Transporter,volcomplete), m_transporter) );
+  }
+}
+
+void
+Carrier::totalvol()
+// *****************************************************************************
+// Sum mesh volumes and contribute own mesh volume to total volume
+// *****************************************************************************
+{
+  // Combine own and communicated contributions of nodal volumes
+  for (const auto& b : m_bid) {
+    auto lid = tk::cref_find( m_lid, b.first );
+    m_vol[ lid ] += m_volc[ b.second ];
+  }
+
+  // Sum mesh volume to host
+  tk::real V = 0.0;
+  for (auto v : m_v) V += v;
+  contribute( sizeof(tk::real), &V, CkReduction::sum_double,
+    CkCallback(CkReductionTarget(Transporter,totalvol), m_transporter) );
+}
+
+void
 Carrier::stat()
```

Comment:
The general idea is to not need deep knowledge of the entire system to understand a block of code.

In this case, let's assume we output data in a manner similar to the following (even if we don't, this is a thought exercise):

```
void printStats()
{
foo.stat();
bar.stat();
man.stat();
}
```

Now, that means nothing. If we go one step better we would have

```
void printStats()
{
inciterArray.stat();
linksysArray.stat();
partitionerArray.stat();
}
```

That is a lot better. Now we know what kinds of objects we are getting data from. So if we understand what kind of data inciter, partitioner, and linksys generate, we are golden.

But why stop there?

```
void printStats()
{
inciterArray.printCellStatistics();
linksysArray.printHypreStatistics();
partitionerArray.printStandardDeviationOfPartitionSize();
}
```

Or something similar. I don't have the deep understanding of quinoa that you do so I am kind of reaching based upon my vague understandings of what those chares did the last time I did a deep dive.

I can't speak to its overall value but I was told to read The Practice of Programming by Kernighan and Pike and it seems pretty useful for explaining methodologies for general coding practices. There are obviously much better books out there, but this one is short and full of examples.